fairlearn/_input_validation.py

```diff
@@ -5,70 +5,133 @@
 
-    # Extract the target attribute
-    sensitive_features_vector = _make_vector(kwargs[_KW_SENSITIVE_FEATURES],
-                                             _KW_SENSITIVE_FEATURES)
+    if type(sensitive_features) not in _ALLOWED_INPUT_TYPES_SENSITIVE_FEATURES:
+        raise TypeError(_INPUT_DATA_FORMAT_ERROR_MESSAGE
+                        .format(_KW_SENSITIVE_FEATURES, _ALLOWED_INPUT_TYPES_SENSITIVE_FEATURES,
+                                type(sensitive_features).__name__))
+
+    if expect_y and type(y) not in _ALLOWED_INPUT_TYPES_Y:
+        raise TypeError(_INPUT_DATA_FORMAT_ERROR_MESSAGE
+                        .format(y.__name__, _ALLOWED_INPUT_TYPES_Y, type(y).__name__))
+
+    if len(X) == 0 or len(sensitive_features) == 0 or (expect_y and len(y) == 0):
+        raise ValueError(_EMPTY_INPUT_ERROR_MESSAGE)
+
+    # Extract the sensitive features. If there are multiple sensitive features compress them into
+    # a single column which will serve as the binning vector.
+    sensitive_features_vector = _make_vector(sensitive_features,
+                                             _KW_SENSITIVE_FEATURES,
+                                             compress_multiple_columns=True)
 
     if enforce_binary_sensitive_feature:
-        unique_labels = np.unique(sensitive_features_vector)
-        if len(unique_labels) > 2:
-            raise RuntimeError("Sensitive features contain more than two unique values")
+        if len(np.unique(sensitive_features_vector)) > 2:
+            raise ValueError(_SENSITIVE_FEATURES_NON_BINARY_ERROR_MESSAGE)
 
-    # Extract the Y values
-    y_vector = _make_vector(y, "y")
+    if expect_y and enforce_binary_labels:
+        if not set(np.unique(y)).issubset(set([0, 1])):
+            raise ValueError(_LABELS_NOT_0_1_ERROR_MESSAGE)
 
     X_rows, _ = _get_matrix_shape(X, "X")
-    if X_rows != y_vector.shape[0]:
-        raise RuntimeError(_MESSAGE_X_Y_ROWS)
+
+    if expect_y:
+        y_vector = _make_vector(y, "y")
+        if X_rows != y_vector.shape[0]:
+            raise ValueError(_MESSAGE_X_Y_ROWS)
+    else:
+        y_vector = None
+
     if X_rows != sensitive_features_vector.shape[0]:
-        raise RuntimeError(_MESSAGE_X_SENSITIVE_ROWS)
+        raise ValueError(_MESSAGE_X_SENSITIVE_ROWS)
 
     return pd.DataFrame(X), y_vector, sensitive_features_vector
 
 
-def _make_vector(formless, formless_name):
+def _make_vector(formless, formless_name, compress_multiple_columns=False):
     formed_vector = None
     if isinstance(formless, list):
         formed_vector = pd.Series(formless)
     elif isinstance(formless, pd.DataFrame):
         if len(formless.columns) == 1:
             formed_vector = formless.iloc[:, 0]
+        elif compress_multiple_columns:
+            formed_vector = formless.apply(
+                lambda row: "-".join(
+                    [str(row[col]).replace("-", "_") for col in formless.columns]),
+                axis=1)
```

Comment:
The use of `replace` can map different values to the same value. My suggestion would be to avoid replace and use a standard separator, either "," or ";". If you're paranoid, you could throw a warning if the separator appears within sensitive features.

Reply:
Also--we should generate the new column name using the same mechanism, e.g., we should end up with column named `race,gender` containing values like `black,male`.

Reply:
Yes, but if they used comma or semicolon in a feature name then you can also map to an already existing feature name. It seems unlikely enough and at least is has a strategy for dealing with such characters. Even if we want comma or semicolon it should be
`replace(';',',')` followed by `';'.join`

Reply:
To wrap this up: in the replace phase, let's use the escape character "\", so:
* replace "\" -> "\\"
* replace "," -> "\,"
(in that order)
After that we can join with "," and all the original information is preserved.

Reply:
Made this change. The column name replacement has not happened because we lose column names through the sklearn input validation. I'll see what we can do about that.

Reply:
reasonable. re. joining of the column names, we should think where/whether we need the joined column names. for the time being, i'm not sure we do. if they get surfaced in the dashboard, the naming is probably handled separately anyway.

Reply:
I agree. This can be addressed if that need arises